parity-clib/examples/cpp/main.cpp

```diff
@@ -16,22 +16,23 @@
 
 #include <chrono>
 #include <parity.h>
+#include <parity.hpp>
 #include <regex>
 #include <string>
 #include <thread>
 
-void* parity_run(std::vector<const char*>);
-int parity_subscribe_to_websocket(void*);
-int parity_rpc_queries(void*);
+static parity_ethereum *parity_run(std::vector<const char *>);
```

Comment:
We should wrap the `parity_ethereum` type provided by the C interface as a C++ class.
There we can define a destructor that does all the required clean-up through calls to the C-API.
This way we can simply return a value-type of the C++ `parity_ethereum` type.
Also for a C++ interface we should generally avoid `std::vector` as input parameters and instead prefer having an iterator (or ranges in C++20) based API. Also there is the `std::string_view` type in C++17 that you might want to use instead of `const char*`. It is quite similar to Rust's `&str`.

If you want to return an optional instance (because I saw returns of `nullptr`) then you should instead return `std::optional<parity_ethereum>` with the C++ wrapper.
If `parity_ethereum` is a sub-class and you want to return a base-class of it, then prefer returning `std::unique_ptr<parity_ethereum>`.

Reply:
The RAII part is done.  I chose to throw C++ exceptions instead of using `optional` for return values.